clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1426,8 +1425,9 @@ private void ensureNotClosed() {
     private void acquire() {
         ensureNotClosed();
         long threadId = Thread.currentThread().getId();
-        if (threadId != currentThread.get() && !currentThread.compareAndSet(NO_CURRENT_THREAD, threadId))
-            throw new ConcurrentModificationException("KafkaConsumer is not safe for multi-threaded access");
+
+        if (!currentThread.compareAndSet(NO_CURRENT_THREAD, Thread.currentThread()) && currentThread.get().getId() != Thread.currentThread().getId())
```

Comment:
@ewencp Thank You for providing comments.

Can you have a look in to below code, and provide your thoughts.

And to this question of yours The problem is that you can't be guaranteed you'll perfectly capture the thread that was in conflict because compareAndSet doesn't let you know what the value was if it wasn't the expected value.

Because compareAndSet fails with expected NO_CURRENT_THREAD, that means the other thread holding the value is nothing but the currentThread.get() value it self right. Please tell me if miss anything?
 private void acquire() {
        ensureNotClosed();
        long accessingThreadId;
        Thread accessingThread = currentThread.get();

        if (accessingThread == null)
            accessingThreadId = -1;
        else
            accessingThreadId = accessingThread.getId();
        
        if(accessingThreadId != Thread.currentThread().getId() && !currentThread.compareAndSet(NO_CURRENT_THREAD,Thread.currentThread()))
            throw new ConcurrentModificationException("KafkaConsumer is not safe for multi-threaded access. Request accessing thread is " + Thread.currentThread().getName() + " and it is already being accessed by " + accessingThread.getName());

        refcount.incrementAndGet();
    }